File: examples/suno_api_examples.ts
/**
 * Suno API Examples - TypeScript/JavaScript
 * 
 * Complete implementation examples for all Suno API v1 endpoints.
 * Replace YOUR_API_KEY with your actual API token.
 * 
 * Requirements:
 * - Node.js 16+ or modern browser with fetch API
 * - TypeScript (optional, but recommended)
 */

const API_BASE_URL = 'https://api.sunoapi.org/api/v1'
const API_KEY = 'YOUR_API_KEY' // Replace with your actual token

const headers = {
  'Authorization': `Bearer ${API_KEY}`,
  'Content-Type': 'application/json'
}

// ========== Core Generation Endpoints ==========

/**
 * 1. Generate Music
 * POST /generate
 */
export async function generateMusic(params: {
  prompt: string
  customMode: boolean
  instrumental?: boolean
  style?: string
  title?: string
  model?: string
  callBackUrl: string
  negativeTags?: string
  vocalGender?: 'm' | 'f'
  styleWeight?: number
  weirdnessConstraint?: number
  audioWeight?: number
  personaId?: string
}) {
  const response = await fetch(`${API_BASE_URL}/generate`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

/**
 * 2. Extend Music
 * POST /generate/extend
 */
export async function extendMusic(params: {
  audioId: string
  defaultParamFlag: boolean
  model: string
  callBackUrl: string
  prompt?: string
  style?: string
  title?: string
  continueAt?: number
  negativeTags?: string
}) {
  const response = await fetch(`${API_BASE_URL}/generate/extend`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

/**
 * 3. Generate Lyrics Only
 * POST /lyrics
 */
export async function generateLyrics(params: {
  prompt: string
  callBackUrl: string
}) {
  const response = await fetch(`${API_BASE_URL}/lyrics`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

// ========== Upload & Transform Endpoints ==========

/**
 * 4. Upload & Cover
 * POST /generate/upload-cover
 */
export async function uploadCover(params: {
  uploadUrl: string
  customMode: boolean
  callBackUrl: string
  prompt?: string
  style?: string
  title?: string
  instrumental?: boolean
  model?: string
}) {
  const response = await fetch(`${API_BASE_URL}/generate/upload-cover`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

/**
 * 5. Upload & Extend
 * POST /generate/upload-extend
 */
export async function uploadExtend(params: {
  uploadUrl: string
  defaultParamFlag: boolean
  callBackUrl: string
  prompt?: string
  style?: string
  title?: string
  continueAt?: number
  model?: string
}) {
  const response = await fetch(`${API_BASE_URL}/generate/upload-extend`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

/**
 * 6. Add Instrumental
 * POST /generate/add-instrumental
 */
export async function addInstrumental(params: {
  uploadUrl: string
  title: string
  tags: string
  callBackUrl: string
  negativeTags?: string
  model?: 'V4_5PLUS' | 'V5'
}) {
  const response = await fetch(`${API_BASE_URL}/generate/add-instrumental`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

/**
 * 7. Add Vocals
 * POST /generate/add-vocals
 */
export async function addVocals(params: {
  uploadUrl: string
  prompt: string
  title: string
  style: string
  callBackUrl: string
  negativeTags?: string
  model?: 'V4_5PLUS' | 'V5'
}) {
  const response = await fetch(`${API_BASE_URL}/generate/add-vocals`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

// ========== Status & Info Endpoints ==========

/**
 * 8. Get Task Status
 * GET /generate/record-info
 */
export async function getTaskStatus(taskId: string) {
  const response = await fetch(
    `${API_BASE_URL}/generate/record-info?taskId=${taskId}`,
    { headers }
  )
  return response.json()
}

/**
 * 9. Get Lyrics Status
 * GET /lyrics/record-info
 */
export async function getLyricsStatus(taskId: string) {
  const response = await fetch(
    `${API_BASE_URL}/lyrics/record-info?taskId=${taskId}`,
    { headers }
  )
  return response.json()
}

/**
 * 10. Get Credits
 * GET /generate/credit
 */
export async function getCredits() {
  const response = await fetch(`${API_BASE_URL}/generate/credit`, { headers })
  return response.json()
}

// ========== Advanced Features ==========

/**
 * 11. Get Timestamped Lyrics (Karaoke)
 * POST /generate/get-timestamped-lyrics
 */
export async function getTimestampedLyrics(params: {
  taskId: string
  audioId?: string
  musicIndex?: number
}) {
  const response = await fetch(
    `${API_BASE_URL}/generate/get-timestamped-lyrics`,
    {
      method: 'POST',
      headers,
      body: JSON.stringify(params)
    }
  )
  return response.json()
}

/**
 * 12. Separate Stems (Vocal/Instrumental)
 * POST /vocal-removal/generate
 */
export async function separateStems(params: {
  taskId: string
  audioId: string
  type: 'separate_vocal' | 'split_stem'
  callBackUrl: string
}) {
  const response = await fetch(`${API_BASE_URL}/vocal-removal/generate`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

/**
 * 13. Get Stem Status
 * GET /vocal-removal/record-info
 */
export async function getStemStatus(taskId: string) {
  const response = await fetch(
    `${API_BASE_URL}/vocal-removal/record-info?taskId=${taskId}`,
    { headers }
  )
  return response.json()
}

/**
 * 14. Convert to WAV
 * POST /wav/generate
 */
export async function convertToWav(params: {
  taskId: string
  audioId: string
  callBackUrl: string
}) {
  const response = await fetch(`${API_BASE_URL}/wav/generate`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

/**
 * 15. Get WAV Status
 * GET /wav/record-info
 */
export async function getWavStatus(taskId: string) {
  const response = await fetch(
    `${API_BASE_URL}/wav/record-info?taskId=${taskId}`,
    { headers }
  )
  return response.json()
}

/**
 * 16. Generate Music Video
 * POST /mp4/generate
 */
export async function generateMusicVideo(params: {
  taskId: string
  audioId: string
  callBackUrl: string
  author?: string
  domainName?: string
}) {
  const response = await fetch(`${API_BASE_URL}/mp4/generate`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

/**
 * 17. Get Video Status
 * GET /mp4/record-info
 */
export async function getVideoStatus(taskId: string) {
  const response = await fetch(
    `${API_BASE_URL}/mp4/record-info?taskId=${taskId}`,
    { headers }
  )
  return response.json()
}

/**
 * 18. Generate Cover Art
 * POST /suno/cover/generate
 */
export async function generateCover(params: {
  taskId: string
  callBackUrl: string
}) {
  const response = await fetch(`${API_BASE_URL}/suno/cover/generate`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

/**
 * 19. Get Cover Status
 * GET /suno/cover/record-info
 */
export async function getCoverStatus(taskId: string) {
  const response = await fetch(
    `${API_BASE_URL}/suno/cover/record-info?taskId=${taskId}`,
    { headers }
  )
  return response.json()
}

/**
 * 20. Boost Style (V4_5+)
 * POST /style/generate
 */
export async function boostStyle(content: string) {
  const response = await fetch(`${API_BASE_URL}/style/generate`, {
    method: 'POST',
    headers,
    body: JSON.stringify({ content })
  })
  return response.json()
}

/**
 * 21. Generate Persona
 * POST /generate/generate-persona
 */
export async function generatePersona(params: {
  taskId: string
  musicIndex: number
  name: string
  description: string
}) {
  const response = await fetch(`${API_BASE_URL}/generate/generate-persona`, {
    method: 'POST',
    headers,
    body: JSON.stringify(params)
  })
  return response.json()
}

// ========== File Upload Endpoints ==========

/**
 * 22. Base64 Upload
 * POST /file-base64-upload
 */
export async function uploadBase64(params: {
  base64Data: string
  uploadPath: string
  fileName?: string
}) {
  const response = await fetch(
    `${API_BASE_URL.replace('/api/v1', '')}/api/file-base64-upload`,
    {
      method: 'POST',
      headers,
      body: JSON.stringify(params)
    }
  )
  return response.json()
}

/**
 * 23. Stream Upload (Browser)
 * POST /file-stream-upload
 */
export async function uploadFile(file: File, uploadPath: string, fileName?: string) {
  const formData = new FormData()
  formData.append('file', file)
  formData.append('uploadPath', uploadPath)
  if (fileName) formData.append('fileName', fileName)

  const response = await fetch(
    `${API_BASE_URL.replace('/api/v1', '')}/api/file-stream-upload`,
    {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${API_KEY}`
      },
      body: formData
    }
  )
  return response.json()
}

/**
 * 24. URL Upload
 * POST /file-url-upload
 */
export async function uploadFromUrl(params: {
  fileUrl: string
  uploadPath: string
  fileName?: string
}) {
  const response = await fetch(
    `${API_BASE_URL.replace('/api/v1', '')}/api/file-url-upload`,
    {
      method: 'POST',
      headers,
      body: JSON.stringify(params)
    }
  )
  return response.json()
}

// ========== Usage Example ==========

async function example() {
  try {
    // 1. Check credits
    const credits = await getCredits()
    console.log('Available credits:', credits.data)

    // 2. Generate simple music
    const musicResult = await generateMusic({
      prompt: 'A relaxing piano melody about sunset in Portugal',
      customMode: false,
      callBackUrl: 'https://your-domain.com/api/music/callback'
    })
    console.log('Music generation started:', musicResult.data.taskId)

    // 3. Poll for status (alternative to callback)
    const taskId = musicResult.data.taskId
    const status = await getTaskStatus(taskId)
    console.log('Status:', status.data.status)

    // 4. When complete, get audio URL
    if (status.data.status === 'SUCCESS') {
      const audioUrl = status.data.sunoData[0].audioUrl
      console.log('Audio ready:', audioUrl)
    }
  } catch (error) {
    console.error('Error:', error)
  }
}

// Uncomment to run example
// example()

File: examples/suno_api_examples.py
"""
Exemplos de implementação da Suno API em Python.
Este script contém funções para cada endpoint principal da Suno API.
Substitua `YOUR_API_KEY` pelo seu token e ajuste os parâmetros conforme necessário.

Requisitos:
- Instale a biblioteca `requests` (ex.: pip install requests).

As funções retornam o objeto Response da biblioteca requests.
"""
import requests
from typing import Dict, Any, Optional

# Configuração básica
API_BASE_URL = "https://api.sunoapi.org/api/v1"
API_KEY = "YOUR_API_KEY"  # Substitua pelo seu token real
HEADERS_JSON = {
    "Authorization": f"Bearer {API_KEY}",
    "Content-Type": "application/json"
}
HEADERS_FORM = {
    "Authorization": f"Bearer {API_KEY}"
}

# 1. Geração de música
def generate_music(params: Dict[str, Any]) -> requests.Response:
    """Cria uma música com base nos parâmetros fornecidos.

    Parâmetros esperados:
      - prompt (str)
      - customMode (bool)
      - instrumental (bool)
      - style (str)
      - title (str)
      - model (str)
      - callBackUrl (str), etc.

    Retorna: Response contendo o taskId.
    """
    url = f"{API_BASE_URL}/generate"
    return requests.post(url, json=params, headers=HEADERS_JSON)

# 2. Extensão de música
def extend_music(params: Dict[str, Any]) -> requests.Response:
    """Estende uma música existente.

    Parâmetros obrigatórios:
      - audioId: ID da música a ser estendida
      - defaultParamFlag (bool)
      - se defaultParamFlag=True: forneça prompt, style, title, continueAt, etc.
    """
    url = f"{API_BASE_URL}/generate/extend"
    return requests.post(url, json=params, headers=HEADERS_JSON)

# 3. Upload e cover de áudio
def upload_cover(params: Dict[str, Any]) -> requests.Response:
    """Carrega um áudio e gera um cover.

    Parâmetros:
      - uploadUrl (str): URL do arquivo original
      - customMode (bool)
      - instrumental (bool)
      - prompt/style/title conforme modo
      - model, callBackUrl etc.
    """
    url = f"{API_BASE_URL}/generate/upload-cover"
    return requests.post(url, json=params, headers=HEADERS_JSON)

# 4. Upload e extensão de áudio
def upload_extend(params: Dict[str, Any]) -> requests.Response:
    """Carrega um áudio e o estende.

    Parâmetros:
      - uploadUrl (str)
      - defaultParamFlag (bool)
      - se defaultParamFlag=True: prompt, style, title, continueAt
      - callBackUrl, model etc.
    """
    url = f"{API_BASE_URL}/generate/upload-extend"
    return requests.post(url, json=params, headers=HEADERS_JSON)

# 5. Adicionar instrumental
def add_instrumental(params: Dict[str, Any]) -> requests.Response:
    """Gera um acompanhamento instrumental.

    Parâmetros:
      - uploadUrl (str)
      - title (str)
      - tags (str)
      - negativeTags (str)
      - callBackUrl (str)
      - modelo (opcional)
    """
    url = f"{API_BASE_URL}/generate/add-instrumental"
    return requests.post(url, json=params, headers=HEADERS_JSON)

# 6. Adicionar vocais
def add_vocals(params: Dict[str, Any]) -> requests.Response:
    """Adiciona vocais a um instrumental existente.

    Parâmetros:
      - uploadUrl (str)
      - prompt (str)
      - title (str)
      - style (str)
      - negativeTags (str)
      - callBackUrl (str)
    """
    url = f"{API_BASE_URL}/generate/add-vocals"
    return requests.post(url, json=params, headers=HEADERS_JSON)

# 7. Obter detalhes de uma tarefa (poll)
def get_task_info(task_id: str) -> requests.Response:
    """Recupera informações detalhadas de uma geração de música.

    Args:
      task_id: ID da tarefa
    """
    url = f"{API_BASE_URL}/generate/record-info?taskId={task_id}"
    return requests.get(url, headers=HEADERS_JSON)

# 8. Obter letras com timestamps
def get_timestamped_lyrics(params: Dict[str, Any]) -> requests.Response:
    """Obtém letras sincronizadas com tempo.

    Parâmetros:
      - taskId (str)
      - audioId (str) ou musicIndex (int)
    """
    url = f"{API_BASE_URL}/generate/get-timestamped-lyrics"
    return requests.post(url, json=params, headers=HEADERS_JSON)

# 9. Potenciar estilo musical
def boost_music_style(content: str) -> requests.Response:
    """Gera um estilo potenciado (modelos V4_5+)."""
    url = f"{API_BASE_URL}/style/generate"
    payload = {"content": content}
    return requests.post(url, json=payload, headers=HEADERS_JSON)

# 10. Gerar persona
def generate_persona(params: Dict[str, Any]) -> requests.Response:
    """Cria uma persona personalizada.

    Parâmetros:
      - taskId (str)
      - musicIndex (int)
      - name (str)
      - description (str)
    """
    url = f"{API_BASE_URL}/generate/generate-persona"
    return requests.post(url, json=params, headers=HEADERS_JSON)

# 11. Gerar capa de música
def generate_music_cover(task_id: str, callback_url: str) -> requests.Response:
    """Solicita geração de capa.

    Args:
      task_id: ID da música original
      callback_url: URL para receber o retorno
    """
    url = f"{API_BASE_URL}/suno/cover/generate"
    payload = {"taskId": task_id, "callBackUrl": callback_url}
    return requests.post(url, json=payload, headers=HEADERS_JSON)

# 12. Obter detalhes da capa
def get_music_cover_info(task_id: str) -> requests.Response:
    url = f"{API_BASE_URL}/suno/cover/record-info?taskId={task_id}"
    return requests.get(url, headers=HEADERS_JSON)

# 13. Gerar letras
def generate_lyrics(prompt: str, callback_url: str) -> requests.Response:
    url = f"{API_BASE_URL}/lyrics"
    payload = {"prompt": prompt, "callBackUrl": callback_url}
    return requests.post(url, json=payload, headers=HEADERS_JSON)

# 14. Obter detalhes das letras
def get_lyrics_info(task_id: str) -> requests.Response:
    url = f"{API_BASE_URL}/lyrics/record-info?taskId={task_id}"
    return requests.get(url, headers=HEADERS_JSON)

# 15. Converter para WAV
def convert_to_wav(task_id: str, audio_id: str, callback_url: str) -> requests.Response:
    url = f"{API_BASE_URL}/wav/generate"
    payload = {"taskId": task_id, "audioId": audio_id, "callBackUrl": callback_url}
    return requests.post(url, json=payload, headers=HEADERS_JSON)

# 16. Obter detalhes do WAV
def get_wav_info(task_id: str) -> requests.Response:
    url = f"{API_BASE_URL}/wav/record-info?taskId={task_id}"
    return requests.get(url, headers=HEADERS_JSON)

# 17. Separação de vocais / stems
def vocal_removal(params: Dict[str, Any]) -> requests.Response:
    """Separa vocal/instrumento ou múltiplos stems.

    Parâmetros:
      - taskId (str)
      - audioId (str)
      - type (str): 'separate_vocal' ou 'split_stem'
      - callBackUrl (str)
    """
    url = f"{API_BASE_URL}/vocal-removal/generate"
    return requests.post(url, json=params, headers=HEADERS_JSON)

# 18. Obter detalhes da separação
def get_vocal_removal_info(task_id: str) -> requests.Response:
    url = f"{API_BASE_URL}/vocal-removal/record-info?taskId={task_id}"
    return requests.get(url, headers=HEADERS_JSON)

# 19. Criar vídeo musical
def create_music_video(task_id: str, audio_id: str, callback_url: str, author: Optional[str] = None, domain_name: Optional[str] = None) -> requests.Response:
    url = f"{API_BASE_URL}/mp4/generate"
    payload = {
        "taskId": task_id,
        "audioId": audio_id,
        "callBackUrl": callback_url
    }
    if author:
        payload["author"] = author
    if domain_name:
        payload["domainName"] = domain_name
    return requests.post(url, json=payload, headers=HEADERS_JSON)

# 20. Obter detalhes do vídeo
def get_music_video_info(task_id: str) -> requests.Response:
    url = f"{API_BASE_URL}/mp4/record-info?taskId={task_id}"
    return requests.get(url, headers=HEADERS_JSON)

# 21. Consultar saldo de créditos
def get_remaining_credits() -> requests.Response:
    url = f"{API_BASE_URL}/generate/credit"
    return requests.get(url, headers=HEADERS_JSON)

# 22. Upload Base64
def base64_upload(base64_data: str, upload_path: str, file_name: Optional[str] = None) -> requests.Response:
    url = f"{API_BASE_URL.replace('/api/v1', '')}/api/file-base64-upload"
    payload = {
        "base64Data": base64_data,
        "uploadPath": upload_path
    }
    if file_name:
        payload["fileName"] = file_name
    return requests.post(url, json=payload, headers=HEADERS_JSON)

# 23. Upload via stream
def stream_upload(file_path: str, upload_path: str, file_name: Optional[str] = None) -> requests.Response:
    url = f"{API_BASE_URL.replace('/api/v1', '')}/api/file-stream-upload"
    files = {"file": open(file_path, "rb")}
    data = {
        "uploadPath": upload_path
    }
    if file_name:
        data["fileName"] = file_name
    return requests.post(url, files=files, data=data, headers=HEADERS_FORM)

# 24. Upload via URL
def url_upload(file_url: str, upload_path: str, file_name: Optional[str] = None) -> requests.Response:
    url = f"{API_BASE_URL.replace('/api/v1', '')}/api/file-url-upload"
    payload = {
        "fileUrl": file_url,
        "uploadPath": upload_path
    }
    if file_name:
        payload["fileName"] = file_name
    return requests.post(url, json=payload, headers=HEADERS_JSON)

if __name__ == "__main__":
    # Exemplo de uso: gerar música simples (modo não-personalizado)
    params = {
        "prompt": "Uma música calma e relaxante de piano.",
        "customMode": False,
        "callBackUrl": "https://seu-dominio.com/callback"  # OBRIGATÓRIO!
    }
    response = generate_music(params)
    print("Status: ", response.status_code)
    print(response.json())
